File: nice-traffic-light/r/R/risk_api_client.R
# Call Evagene's /risk/calculate endpoint for the NICE model.

HTTP_OK_LOWER <- 200L
HTTP_OK_UPPER <- 300L

#' Raise an api_error condition.
#' @noRd
stop_api <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("api_error", "error", "condition")
  ))
}

#' Call the Evagene risk/calculate endpoint with model=NICE.
#'
#' @param gateway A gateway function built by e.g. `make_httr2_gateway()`.
#' @param base_url Evagene base URL (trailing slash optional).
#' @param api_key Value for the `X-API-Key` header.
#' @param pedigree_id UUID string.
#' @param counselee_id Optional UUID string.
#' @return A named list parsed from the response JSON.
calculate_nice <- function(gateway, base_url, api_key, pedigree_id, counselee_id = NULL) {
  url <- sprintf("%s/api/pedigrees/%s/risk/calculate", sub("/$", "", base_url), pedigree_id)
  body <- list(model = "NICE")
  if (!is.null(counselee_id)) {
    body$counselee_id <- counselee_id
  }

  response <- gateway(list(
    url = url,
    headers = c("X-API-Key" = api_key, "Accept" = "application/json"),
    body_json = jsonlite::toJSON(body, auto_unbox = TRUE)
  ))

  if (response$status < HTTP_OK_LOWER || response$status >= HTTP_OK_UPPER) {
    stop_api(sprintf("Evagene API returned HTTP %d for %s", response$status, url))
  }

  parsed <- tryCatch(
    jsonlite::fromJSON(response$json_body, simplifyVector = FALSE),
    error = function(e) {
      stop_api(sprintf("Evagene API returned invalid JSON: %s", conditionMessage(e)))
    }
  )
  if (!is.list(parsed) || is.null(names(parsed))) {
    stop_api("Evagene API returned non-object JSON")
  }
  parsed
}

File: nice-traffic-light/node/src/riskApiClient.ts
import type { HttpGateway } from './httpGateway.js';

const HTTP_OK_LOWER = 200;
const HTTP_OK_UPPER = 300;

export class ApiError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ApiError';
  }
}

export interface RiskApiClientOptions {
  readonly baseUrl: string;
  readonly apiKey: string;
  readonly http: HttpGateway;
}

export interface CalculateNiceArgs {
  readonly pedigreeId: string;
  readonly counseleeId?: string;
}

export class RiskApiClient {
  private readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly http: HttpGateway;

  constructor(options: RiskApiClientOptions) {
    this.baseUrl = options.baseUrl.replace(/\/$/, '');
    this.apiKey = options.apiKey;
    this.http = options.http;
  }

  async calculateNice(args: CalculateNiceArgs): Promise<unknown> {
    const url = `${this.baseUrl}/api/pedigrees/${args.pedigreeId}/risk/calculate`;
    const body: Record<string, unknown> = { model: 'NICE' };
    if (args.counseleeId !== undefined) {
      body.counselee_id = args.counseleeId;
    }

    const response = await this.http.postJson(url, { headers: this.headers(), body });
    if (response.status < HTTP_OK_LOWER || response.status >= HTTP_OK_UPPER) {
      throw new ApiError(`Evagene API returned HTTP ${response.status.toString()} for ${url}`);
    }
    return response.json();
  }

  private headers(): Record<string, string> {
    return {
      'X-API-Key': this.apiKey,
      'Content-Type': 'application/json',
      Accept: 'application/json',
    };
  }
}
